### DISCUSSION TITLE
Trigger performance using SqlServer persistent store

### DISCUSSION THREAD

**Discussion by alexdawes** [quartznet/quartznet#1589] [Q&A]
Hi,

I've been experimenting with Quartz, using the SqlServer ADO-based persistent store. My simple scheduler setup looks something like this:

```c#
            return await SchedulerBuilder.Create()
                .WithId(Guid.NewGuid().ToString())
                .UseDefaultThreadPool(x => x.MaxConcurrency = 100)
                .UsePersistentStore(x =>
                {
                    x.UseProperties = true;
                    x.UseClustering();
                    x.UseSqlServer(c =>
                    {
                        c.ConnectionString =
                                "Server=[redacted];Database=[redacted];User Id=[redacted];Password=[redacted];Max Pool Size=100;TrustServerCertificate=true;";
                    });
                    x.UseJsonSerializer();
                })
                .BuildScheduler();
```

I've noticed that when triggers are firing, they seemed to be capped to about 1 trigger per 1-3 seconds - and this is regardless of how many nodes I run in a cluster.

As an example, I am creating 10 jobs, each of which is scheduled to trigger every 10 seconds:

```c#
            for (var i = 0; i < 10; i++)
            {
                var job = JobBuilder.Create<HelloJob>()
                    .WithIdentity($"Job-{i+1}")
                    .RequestRecovery()
                    .Build();
            
                var trigger = TriggerBuilder.Create()
                    .WithIdentity($"Trigger-{i+1}")
                    .ForJob(job)
                    .StartNow()
                    .WithCronSchedule("0/10 * * * * ?", b => { b.WithMisfireHandlingInstructionIgnoreMisfires(); })
                    .Build();
                
                await scheduler.ScheduleJob(job, trigger);
            }
```

I'm using `WithMisfireHandlingInstructionIgnoreMisfires` because I need all misfires to be eventually executed. The `HelloJob` is doing a simple print statement via the console.

```c#
    public class HelloJob : IJob
    {
        public async Task Execute(IJobExecutionContext context)
        {
            await Console.Out.WriteLineAsync(
                $"Job {context.JobDetail.Key.Name} triggered at {context.ScheduledFireTimeUtc}, executed at {context.FireTimeUtc}");
        }
    }
```

What I've observed is that when the triggers fire, they appear to fire one at a time, and the process handling this can only handle about 1 per second. Because of this, my test using 10 jobs triggered every 10 seconds hits the limit, and the scheduler is simply not able to keep up - the gap between the scheduled fire time and the actual fire time begins getting larger and larger...

```
Job Job-2 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:33:58 +00:00
Job Job-3 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:33:59 +00:00
Job Job-1 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:01 +00:00
Job Job-4 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:02 +00:00
Job Job-5 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:04 +00:00
Job Job-6 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:05 +00:00
Job Job-7 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:06 +00:00
Job Job-8 triggered at 04/22/2022 14:33:50 +00:00, executed at 04/22/2022 14:34:08 +00:00
Job Job-1 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:09 +00:00
Job Job-10 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:11 +00:00
Job Job-2 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:12 +00:00
Job Job-3 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:14 +00:00
Job Job-4 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:15 +00:00
Job Job-5 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:17 +00:00
Job Job-6 triggered at 04/22/2022 14:34:00 +00:00, executed at 04/22/2022 14:34:18 +00:00
```

The slowness is definitely in the SqlServer job store and not on the C# side - when I switch to a RAM-based store, all 10 triggers fire together and immediately (note that the execution timestamp matches the trigger timestamp):

```
Job Job-2 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-4 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-7 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-6 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-3 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-9 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-1 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-10 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-5 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-8 triggered at 04/22/2022 14:35:40 +00:00, executed at 04/22/2022 14:35:40 +00:00
Job Job-1 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-10 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-2 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-3 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-4 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-5 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-6 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-7 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-8 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
Job Job-9 triggered at 04/22/2022 14:35:50 +00:00, executed at 04/22/2022 14:35:50 +00:00
```

I'm concerned about the ability to scale here, as it seems that having an average number of triggers above 1 per second will mean that the scheduler is unable to keep up with the triggers, and will instead create a huge backlog. Given this, I'm not confident I will be able to use Quartz at all.

My database here is a part of a dev environment, but should be sufficiently powerful. 

**Is there anything else I can do to get better performance when using the Sql Server job store?**

**Do any other persistent job stores offer better performance?**

**Comment by alexdawes** [comment]
I should add, spinning up additional processes with new schedulers to try to handle some of the load does not seem to have any impact on the total number of triggers per second.

**Comment by alexdawes** [comment]
Update - I switched from using our dev database to one running locally, and the performance is ALOT better. Looks like latency between my machine and the DB was the culprit. Hopefully performance will be fine when running on a cloud hosted kube cluster instead of locally, as the latency should be much smaller.

**Comment by lahma** [comment]
I did check the example code I couldn't spot anything obvious. Like you've stated, it might be the latency. There are things to tune like how many triggers to acquire at the same time but nothing would explain why there wouldn't be concurrent invocations of jobs which are not marked with `[DisallowConcurrentExecution]`.

**Comment by alexdawes** [comment]
Jobs can execute concurrently - when I make the jobs long running, I can get multiple jobs running at once. However, it seems to be the process that is delegating the jobs across the cluster and/or starting them thats slow. For now, I can work under the assumption that its latency thats causing the problem, and hopefully can test it again when I get to the point of deploying my service into the cloud.
